README.md
# fault-tolerant-jsonrpc [![CircleCI](https://circleci.com/gh/segmentio/fault-tolerant-jsonrpc.svg?style=svg&circle-token=9924e7270d858226e8489267037e8d363c6af5ff)](https://circleci.com/gh/segmentio/fault-tolerant-jsonrpc)

> Fault tolerant jsonrpc with retries and timeouts

This module provides fault-tolerance on top of [@segment/jsonrpc2](github.com/segmentio/jsonrpc2.js).

- [x] Retries via [`p-retry`](https://github.com/sindresorhus/p-retry)
- [x] Timeout via [`p-timeout`](https://github.com/sindresorhus/p-timeout)
- [x] Optional retry logic
- [ ] Circuit Breaking not yet implemented


#### Note about retries

Retries are **off by default** and should only be used for idempotent RPC methods.

## Install

```
$ npm install --save @segment/fault-tolerant-jsonrpc
```

## Usage

```js
const jsonrpc = require('@segment/fault-tolerant-jsonrpc')

// This is an example service client,
// implemented using fault-tolerant-jsonrpc
function Client (addr, opts) {
  opts = opts || {}

  // We should only allow retries for idempotent requests
  const idempotentDefaults = Object.assign({
    retryOptions: { retries: 3 },

    // Custom retry function
    // This allows the clients to decide what sorts of errors
    // are worth retrying
    retryCondition: function(originalError) {
      if (originalError.code === "SYSTEM_ERROR") {
        return true
      }
      return false
    }
    timeout: 500,
    totalTimeout: 2000
  }, opts.idempotentDefaults)

  const rpc = jsonrpc(addr, opts.globalDefaults)

  return {
    getAll () {
      return rpc.call('Items.GetAll', null, idempotentDefaults)
    },
    getOne (id) {
      return rpc.call('Items.GetOne', id, idempotentDefaults)
    },
    createOne (data) {
      // Intentionally not using retries since this
      // RPC method _could_ result in duplicate writes
      return rpc.call('Items.CreateOne', data)
    }
  }
}
```

``` js
const itemService = Client('http://localhost:3000')

itemService
  .getAll()
  .then(console.log)
  .catch(console.log)

itemService
  .getOne('item-1')
  .then(console.log)
  .catch(console.log)

itemService
  .createOne({ some: 'thing' })
  .then(console.log)
  .catch(console.log)
```

## API

### `const rpc = jsonrpc(addr, [options])`

Returns a jsonrpc client where `client.call` supports retries and timeouts. This
package exposes the same API as [jsonrpc2](https://github.com/segmentio/jsonrpc2.js)
with the addition of the following options:

- `options.retryOptions` is passed to the [`p-retry`](https://github.com/sindresorhus/p-retry#options) module.
- `options.totalTimeout` is passed to the [`p-timeout`](https://github.com/sindresorhus/p-timeout) module.

### `rpc.call(method, params, [options])`

Same API as the above but global options can be overwritten on a per request basis.


History.md
1.3.1 / 2018-02-07
==================

  * Update license and package.json

1.3.0 / 2017-05-04
==================

  * Allow clients to provide a retryCondition function (#2)
  * Return descriptive error messages when the RPC client times out (#1)


test.js

import test from 'ava'
import nock from 'nock'
import jsonrpc, { RPCTimeoutError } from './'

const HOST = 'http://test.rpc'
const rpc = jsonrpc(`${HOST}/rpc`)

test.serial('success', async t => {
  nock(HOST)
    .post('/rpc', {
      method: 'Items.GetOne',
      params: [1],
      id: /.+/,
      jsonrpc: '2.0'
    })
    .reply(200, {
      result: { el: 'duderino' }
    })

  const result = await rpc.call('Items.GetOne', 1)
  const expected = { el: 'duderino' }

  t.deepEqual(result, expected)
})

test.serial('retry', async t => {
  let i = 0
  nock(HOST)
    .filteringRequestBody(() => '*')
    .post('/rpc', '*')
    .times(3)
    .reply(200, () => {
      i++
      return { error: 'failed' }
    })

  const options = {
    retryOptions: {
      retries: 2,
      minTimeout: 10
    }
  }

  const error = await t.throws(rpc.call('Items.GetAll', null, options))

  t.is(i, 3)
  t.is(error.message, 'failed')
})

test.serial('retry off by default', async t => {
  let i = 0
  nock(HOST)
    .filteringRequestBody(() => '*')
    .post('/rpc', '*')
    .reply(200, () => {
      i++
      return { error: 'failed' }
    })

  const error = await t.throws(rpc.call('Items.GetAll'))

  t.is(i, 1)
  t.is(error.message, 'failed')
})

test.serial('timeout', async t => {
  const timeout = 10
  const retries = 100
  let i = 0

  nock(HOST)
    .filteringRequestBody(() => '*')
    .post('/rpc', '*')
    .times(retries * 2)
    .socketDelay(100)
    .reply(200, () => {
      i++
      return { result: { el: 'duderino' } }
    })

  const options = {
    retryOptions: {
      retries,
      factor: 1,
      minTimeout: timeout,
      maxTimeout: timeout
    },
    timeout: timeout - 5,
    totalTimeout: 10000
  }

  const start = Date.now()
  const error = await t.throws(rpc.call('Items.GetTimeout', null, options))
  const duration = Date.now() - start
  const expectedDuration = (timeout + 3) * retries

  t.truthy(duration > expectedDuration * 0.85)
  t.truthy(duration < expectedDuration * 1.15)
  t.is(i, retries + 1)
  t.is(error.message, 'ESOCKETTIMEDOUT')
})

test.serial('totalTimeout', async t => {
  nock(HOST)
    .filteringRequestBody(() => '*')
    .post('/rpc', '*')
    .times(10)
    .socketDelay(100)
    .reply(200, {
      result: {
        el: 'duderino'
      }
    })

  const options = {
    retryOptions: {
      retries: 20,
      factor: 1,
      minTimeout: 15,
      maxTimeout: 15
    },
    timeout: 10,
    totalTimeout: 100
  }

  const start = Date.now()
  const error = await t.throws(rpc.call('Items.GetTotalTimeout', null, options))
  const duration = Date.now() - start

  t.truthy(duration < 110 && duration > 90)
  t.truthy(error.message.match(/RPC Client Timed out after \d* attempts/))
  t.truthy(error instanceof RPCTimeoutError)
})


index.js
'use strict'

const RPC = require('@segment/jsonrpc2')
const pRetry = require('p-retry')
const pTimeout = require('p-timeout')
const RPCTimeoutError = require('./lib/rpc_timeout_error')

const noRetryDefault = { retries: 0 }
const defaultRetryPredicate = () => true

module.exports = Client
module.exports.RPCTimeoutError = RPCTimeoutError

function Client (addr, libraryOptions) {
  if (!(this instanceof Client)) return new Client(addr, libraryOptions)

  libraryOptions = libraryOptions || {}

  const originalRpcClientOptions = {
    timeout: libraryOptions.timeout,
    logger: libraryOptions.logger
  }

  const rpc = new RPC(addr, originalRpcClientOptions)
  const call = rpc.call.bind(rpc)

  rpc.call = function (method, params, options) {
    options = options || {}

    // Allow overriding options at the individual request level.
    // Retries are opt-in and turned off by default.
    const retryOptions = Object.assign({}, noRetryDefault, libraryOptions.retryOptions, options.retryOptions)
    const totalTimeout = options.totalTimeout || libraryOptions.totalTimeout || 1000

    // Decides whether or not we should retry a given request.
    // The client can decide if they want to retry 404s, or just other types
    // of errors such as ECONNREFUSED, ESOCKETTIMEDOUT and other common network errors
    const retryPredicate = libraryOptions.shouldRetry || options.shouldRetry || defaultRetryPredicate

    return new Promise((resolve, reject) => {
      let hasTimedOut = false
      let currentAttempt = 0

      const run = attempts => {
        currentAttempt = attempts

        if (hasTimedOut) {
          return pRetry.AbortError
        }

        return call(method, params, {
          timeout: options.timeout,
          async: options.async
        }).catch(shouldRetry)
      }

      function shouldRetry (err) {
        if (retryPredicate(err)) {
          // Keep retrying
          return Promise.reject(err)
        } else {
          // Abort retry flow
          throw new pRetry.AbortError(err)
        }
      }

      function fallback () {
        const message = `RPC Client Timed out after ${currentAttempt} attempts`

        const error = new RPCTimeoutError(message, {
          attempts: currentAttempt,
          method,
          params
        })

        return Promise.reject(error)
      }

      pTimeout(pRetry(run, retryOptions), totalTimeout, fallback)
        .then(resolve)
        .catch(err => {
          hasTimedOut = true
          reject(err)
        })
    })
  }

  return rpc
}


example.js
const jsonrpc = require('./')

// This is an example service client,
// implemented using fault-tolerant-jsonrpc
function Client (addr, opts) {
  opts = opts || {}

  // We should only allow retries for idempotent requests
  const idempotentDefaults = Object.assign({
    retryOptions: { retries: 3 },
    timeout: 500,
    totalTimeout: 2000
  }, opts.idempotentDefaults)

  const rpc = jsonrpc(addr, opts.globalDefaults)

  return {
    getAll () {
      return rpc.call('Items.GetAll', null, idempotentDefaults)
    },
    getOne (id) {
      return rpc.call('Items.GetOne', id, idempotentDefaults)
    },
    createOne (data) {
      // Intentionally not using retries since this
      // RPC method _could_ result in duplicate writes
      return rpc.call('Items.CreateOne', data)
    }
  }
}

const itemService = Client('http://localhost:3000')

itemService
  .getAll()
  .then(console.log)
  .catch(console.log)

itemService
  .getOne('item-1')
  .then(console.log)
  .catch(console.log)

itemService
  .createOne({ some: 'thing' })
  .then(console.log)
  .catch(console.log)
